lib/vector-core/src/config/mod.rs

```diff
@@ -100,42 +100,129 @@ impl Input {
     // the relevant schemas, and store them separately in this type.
-    ///
-    /// The `None` variant of a schema definition has two distinct meanings for a source component
-    /// versus a transform component:
-    ///
-    /// For *sources*, a `None` schema is identical to a `Some(Definition::source_default())`.
-    ///
-    /// For a *transform*, a schema [`schema::Definition`] is required if `Datatype` is Log.
-    pub log_schema_definition: Option<schema::Definition>,
+    pub schema_definition: Option<schema::Definition>,
 }
 
-impl Output {
-    /// Create a default `Output` of the given data type.
+impl SourceOutput {
+    /// Create a `SourceOutput` of the given data type that contains a single output `Definition`.
+    /// Designed for use in log sources.
     ///
-    /// A default output is one without a port identifier (i.e. not a named output) and the default
-    /// output consumers will receive if they declare the component itself as an input.
-    pub fn default(ty: DataType) -> Self {
+    /// Asserts that `ty` contains [`DataType::Log`].
+    #[must_use]
+    pub fn new_logs(ty: DataType, schema_definition: schema::Definition) -> Self {
+        assert!(ty.contains(DataType::Log));
+
         Self {
             port: None,
             ty,
-            log_schema_definition: None,
+            schema_definition: Some(schema_definition),
+        }
+    }
+
+    /// Create a `SourceOutput` of the given data type that contains no output `Definition`s.
+    /// Designed for use in metrics sources.
+    ///
+    /// Sets the datatype to be [`DataType::Metric`].
+    #[must_use]
+    pub fn new_metrics() -> Self {
+        Self {
+            port: None,
+            ty: DataType::Metric,
+            schema_definition: None,
+        }
+    }
+
+    /// Create a `SourceOutput` of the given data type that contains no output `Definition`s.
+    /// Designed for use in trace sources.
+    ///
+    ///
+    /// Sets the datatype to be [`DataType::Trace`].
+    #[must_use]
+    pub fn new_traces() -> Self {
+        Self {
+            port: None,
+            ty: DataType::Trace,
+            schema_definition: None,
         }
     }
 
-    /// Set the schema definition for this `Output`.
+    /// Return the schema [`schema::Definition`] from this output.
+    ///
+    /// Takes a `schema_enabled` flag to determine if the full definition including the fields
+    /// and associated types should be returned, or if a simple definition should be returned.
+    /// Schema enabled is set in the users configuration.
     #[must_use]
-    pub fn with_schema_definition(mut self, schema_definition: schema::Definition) -> Self {
-        self.log_schema_definition = Some(schema_definition);
+    pub fn schema_definition(&self, schema_enabled: bool) -> Option<schema::Definition> {
+        self.schema_definition.as_ref().map(|definition| {
+            if schema_enabled || definition.log_namespaces().contains(&LogNamespace::Vector) {
```

Comment:
The `Vector` namespace should only make sure that semantic meanings are included, not the full type definition.

Reply:
So like this?

```
if schema_enabled:
  Full definitions are returned
else if `Vector Namespace`:
  Just return the meanings
else:
  return Nothing
```

Reply:
yes, where "nothing" means the default for the namespace